File: Expe3.py
class Animal:
    def hacer_sonido(self):
        raise NotImplementedError

class Perro(Animal):
    def hacer_sonido(self):
        return "¡Guau!"

class Gato(Animal):
    def hacer_sonido(self):
        return "¡Miau!"

class Vaca(Animal):
    def hacer_sonido(self):
        return "¡Muuu!"

def comprobar_sonido(animal, sonido_esperado):
    if animal.hacer_sonido() == sonido_esperado:
        print(f"¡Correcto! El sonido es: {animal.hacer_sonido()}")
    else:
        print(f"Incorrecto. El sonido es: {animal.hacer_sonido()}")

def main():
    perro = Perro()
    gato = Gato()
    vaca = Vaca()

    while True:
        print("Seleccione un animal para escuchar su sonido:")
        print("1. Perro")
        print("2. Gato")
        print("3. Vaca")
        print("4. Salir")
        opcion = int(input())

        if opcion == 4:
            break

        sonido_esperado = input("Ingrese el sonido esperado: ")

        if opcion == 1:
            comprobar_sonido(perro, sonido_esperado)
        elif opcion == 2:
            comprobar_sonido(gato, sonido_esperado)
        elif opcion == 3:
            comprobar_sonido(vaca, sonido_esperado)
        else:
            print("Opción inválida. Inténtelo de nuevo.")

if __name__ == "__main__":
    main()

File: Expe3.js
class Animal {
    hacerSonido() {
        throw new Error("Este método debe ser implementado por las subclases");
    }
}

class Perro extends Animal {
    hacerSonido() {
        return "¡Guau!";
    }
}

class Gato extends Animal {
    hacerSonido() {
        return "¡Miau!";
    }
}

class Vaca extends Animal {
    hacerSonido() {
        return "¡Muuu!";
    }
}

function comprobarSonido(animal, sonidoEsperado) {
    if (animal.hacerSonido() === sonidoEsperado) {
        console.log("¡Correcto! El sonido es: " + animal.hacerSonido());
    } else {
        console.log("Incorrecto. El sonido es: " + animal.hacerSonido());
    }
}

const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const perro = new Perro();
const gato = new Gato();
const vaca = new Vaca();

function preguntar() {
    console.log("Seleccione un animal para escuchar su sonido:");
    console.log("1. Perro");
    console.log("2. Gato");
    console.log("3. Vaca");
    console.log("4. Salir");

    rl.question('Ingrese una opción: ', (opcion) => {
        if (opcion === '4') {
            rl.close();
            return;
        }

        rl.question('Ingrese el sonido esperado: ', (sonidoEsperado) => {
            switch (opcion) {
                case '1':
                    comprobarSonido(perro, sonidoEsperado);
                    break;
                case '2':
                    comprobarSonido(gato, sonidoEsperado);
                    break;
                case '3':
                    comprobarSonido(vaca, sonidoEsperado);
                    break;
                default:
                    console.log("Opción inválida. Inténtelo de nuevo.");
            }
            preguntar();
        });
    });
}

preguntar();
